### PR TITLE
Tracking Pixel, Shortcode Embed Forms, Shortcode Embed Dynamic Content

### PR DESCRIPTION
**Pull request by badri**
Test with latest Mautic as of now(2.4)

Tracking pixel embed not working, js gives same origin error.

### CODE REVIEW

**Comment by escopecz** [comment]
I'm a bit confused. Your description seems to be an issue report. What are the code changes about?

For the origin error, turn on the CORS setting in your Mautic and white-list your website domain.

**Comment by badri** [comment]
Yes, but it was happening in Mautic cloud as well, so thought it was worth mentioning. If the tracking pixel can be put in place of the javascript without any loss of functionality, as mentioned in the Getting started document, I can make that change and push.

**Comment by escopecz** [comment]
It doesn't matter where the Mautic instance is. CORS is browser security feature and you have to enable it in Mautic configuration if your Mautic and website doesn't run on the same domain.

Could you update the PR description and add what your goal is, what the changes are doing?

**Comment by badri** [comment]
This PR adds the following changes:

1. Add Mautic JS tracker on selected set of pages.
2. Short code for forms.
3. Short code for Dynamic Web content.

**Comment by jordan8037310** [comment]
To shine some light on this:

Under Configuration > Content Authoring > Text Formats

Select the text format that is default on your content types.

Hit the Checkbox for Mautic Shortcodes: 
image

Update the Filter Processing Order as needed:
image

And then you just insert the shortcode on any field which has the Text Input Format processing setup, e.g. `[mauticform id="13"]` or `[mauticform id="ID"]`

This update was very useful for us, and the first implementation for Drupal 7 that actually worked well with form injection. Thank you!

**Comment by jordan8037310** [comment]
@escopecz - Was there anything else to address before merging this in?

**Comment by escopecz** [comment]
I don't see any comment about testing or code reviewing this PR. The core team is focused on Mautic itself at the moment. Would someone from the Drupal community be willing to help maintaining this repository?

**Comment by jordan8037310** [comment]
@escopecz - I'm happy to help maintain. I will do a test of the Focus Items and see if there are any issues there.

For us, this update was required in order to get the Mauticforms working with shortcode replacement. That has been tested and verified per my instructions from 2/1/17.

File: README.md
Comment (jordan8037310) [inline README.md:12]: Misspelling: `.. Save Configuration`.

File: README.md
Comment (jordan8037310) [inline README.md:18]: `[mauticform id="ID"]`

**Review by jordan8037310 — APPROVED** [review]
Some typos in documentation

**Comment by jordan8037310** [comment]
Correction, Focus Items are not included in this pull request. Future Pull Request will be sent to include that functionality.

I cannot verify that Dynamic Content is working. I'm not sure how to go about configuring it locally as I've never used it. @escopecz can you clarify for me? Is the Components > Dynamic Content item supposed to be insertable onto a web page on its own? Or do Dynamic Content "slots" get created only from a campaign creation? Documentation seems to indicate to me that Dynamic Content is only created on Campaign creation.

The code in Drupal works, it replaces the shortcode with the dynamic content, but it is unclear to me if an additional javascript file should load data into this slot: `<div data-slot-container=""><div class="mautic-slot" data-slot-name="@slot_name">@dom</div></div>`

**Comment by escopecz** [comment]
It can be created from both places. For example as emails does.

**Comment by jordan8037310** [comment]
@escopecz - Sorry, those instructions don't really make sense. Can Dynamic Content be injected on third-party landing pages? Are Dynamic Content slots on third-party landing pages only managed by Campaigns? What parameters are passed to a third-party landing page in order for that content slot to be populated with content, is it a url param? 

These questions aren't answered by the documentation, hoping you can shine some light here.

**Comment by escopecz** [comment]
1. Yes, the dynamic content works on 3rd party websites tracked by Mautic JS.
2. The content can be created/updated via campaigns as well as its own menu item. They are triggered by the campaign actions and decisions.
3. The JS searching for the dynamic content slot HTML. The needed JS is already part of the Mautic tracking JS, so no need to add anything new. If the JS finds some dynamic content, it will ask Mautic what content it should display for the current contact and if there is some content, it will replace the default content with the dynamic one.

I didn't develop this feature, maybe @dongilbert will be able to explain it better or correct me.

**Comment by jordan8037310** [comment]
Perfect, thank you @escopecz. I was lost as to how the Slot knew to switch between Dynamic Content variants. I will  test this with a campaign.

**Comment by jordan8037310** [comment]
@dongilbert is data-slot-name supposed to be the ID # for the Dynamic Content? as In 

`<div data-slot-container="content"><div class="mautic-slot" data-slot-name="1">My Default Content</div></div>`

**Comment by jordan8037310** [comment]
I currently can't test this because I can't get Push Dynamic Content to work in my campaign. Campaign is very simple:

image

But no dynamic content pushes are working:
image

even though the condition 100% fires.
image

So this issue is likely blocked by some bug in Mautic. I am on 2.8.0. Will have to see if someone can point me in the direction of a bug fix.

Would really like to get this branch merged in, though. Perhaps we should decouple the dynamic content work from this pull request and re-merge in a separate branch?

**Comment by jordan8037310** [comment]
I am going to merge this Pull Request. The work in the Drupal module correctly calls the dynamic web content endpoint. The point I made earlier about DWC not working in my current version of Mautic still stands, but as far as I've been able to test on the Drupal side the module is operating fine, replacing the shortcode, and making the call to Mautic dwc path.

**Comment by rteijeiro** [comment]
I have the same problem with Dynamic Content, it seems to be an issue with latest version of Mautic: https://github.com/mautic/mautic/issues/4061

**Pull request merged**
